src/Raven.Server/Documents/Indexes/IndexStore.cs

```diff
@@ -1439,6 +1466,87 @@ public IndexMergeResults ProposeIndexMergeSuggestions()
 
             return indexMerger.ProposeIndexMergeSuggestions();
         }
+
+        public class IndexBatchScope
+        {
+            private readonly IndexStore _store;
+
+            private PutIndexesCommand _command;
+
+            public IndexBatchScope(IndexStore store)
+            {
+                _store = store;
+            }
+
+            public void AddIndex(IndexDefinitionBase definition)
+            {
+                if (_command == null)
+                    _command = new PutIndexesCommand(_store._documentDatabase.Name);
+
+                if (definition == null)
+                    throw new ArgumentNullException(nameof(definition));
+
+                if (definition is MapIndexDefinition indexDefinition)
+                {
+                    AddIndex(indexDefinition.IndexDefinition);
+                    return;
+                }
+
+                _store.ValidateAutoIndex(definition);
+
+                var autoDefinition = (AutoIndexDefinitionBase)definition;
+                var indexType = PutAutoIndexCommand.GetAutoIndexType(autoDefinition);
+
+                _command.Auto.Add(PutAutoIndexCommand.GetAutoIndexDefinition(autoDefinition, indexType));
+            }
+
+            public void AddIndex(IndexDefinition definition)
+            {
+                if (_command == null)
+                    _command = new PutIndexesCommand(_store._documentDatabase.Name);
+
+                _store.ValidateStaticIndex(definition);
+
+                _command.Static.Add(definition);
+            }
+
+            public async Task SaveAsync()
+            {
+                if (_command == null || _command.Static.Count == 0 && _command.Auto.Count == 0)
+                    return;
+
+                try
+                {
+                    var (index, _) = await _store._serverStore.SendToLeaderAsync(_command);
```

Comment:
If an exception occurred during `SendToLeaderAsync` it is probably due to network issues / cluster down etc.
Meaning that the command is not committed to the raft log and wouldn't be executed.

Any later exception will mean that the command is committed but it's execution has failed or wasn't completed in time.

Reply:
That's fine. We are only catching TimeoutExeception here. Same flow as it was/is in `CreateIndex`

Reply:
Okay, but if we passed this line then the explanation for `TimeoutExeception` is incorrect, because the cluster is not down :)

Reply:
Oh... that is what you meant. I will fix the exeception. Give me few minutes.

Reply:
Done